File: app/app/src/main/java/com/europad/app/net/Proto.kt
package com.europad.app.net

object Proto {
    const val MAGIC: Int = 0x01E0
    const val MAGIC_LO: Byte = 0xE0.toByte()
    const val MAGIC_HI: Byte = 0x01
    const val VERSION: Byte = 2
    const val FRAME_SIZE = 30
    const val AXES_COUNT = 8
    const val DEFAULT_PORT = 47910

    // Written as literals, not `1 shl n`: a Byte-typed shl expression is an Int today and Kotlin has
    // announced it will stop silently narrowing, which would break the build on a compiler bump.
    const val FLAG_PING_REQUEST: Byte = 0x01
    const val FLAG_PING_REPLY: Byte = 0x02
    const val FLAG_HELLO: Byte = 0x04
    const val FLAG_RUMBLE: Byte = 0x08
    const val FLAG_ACK: Byte = 0x10
    const val FLAG_REJECT: Byte = 0x20
    const val FLAG_CONFIG: Byte = 0x40

    const val CFG_SET_BIT_KEY: Byte = 1

    const val OFF_MAGIC = 0
    const val OFF_VERSION = 2
    const val OFF_FLAGS = 3
    const val OFF_SEQ = 4
    const val OFF_TIMESTAMP = 6
    const val OFF_BUTTONS_LO = 10
    const val OFF_BUTTONS_HI = 12
    const val OFF_AXES = 14

    const val REJECT_VERSION_MISMATCH = 1
    const val REJECT_WRONG_PIN = 2
    const val REJECT_LOBBY_FULL = 3

    object Axis {
        const val LX = 0
        const val LY = 1
        const val RX = 2
        const val RY = 3
        const val LT = 4
        const val RT = 5
        const val STEER = 6
        const val AUX0 = 7
    }
}

File: server/EuroPad.Server/Protocol/Proto.cs
namespace EuroPad.Server.Protocol;

public static class Proto
{
    public const ushort Magic = 0x01E0;
    public const byte Version = 2;
    public const int FrameSize = 30;
    public const int AxesCount = 8;

    public const ushort DefaultPort = 47910;

    public const byte FlagPingRequest = 1 << 0;
    public const byte FlagPingReply = 1 << 1;
    public const byte FlagHello = 1 << 2;
    public const byte FlagRumble = 1 << 3;
    public const byte FlagAck = 1 << 4;
    public const byte FlagReject = 1 << 5;
    public const byte FlagConfig = 1 << 6;

    public const byte CfgSetBitKey = 1;

    /// <summary>
    /// Every flag that marks a frame as control-only. Such frames carry no axis/button payload, so
    /// they must never reach the pad — see <see cref="InputFrame.CarriesInput"/>.
    /// </summary>
    public const byte ControlFlags =
        FlagPingRequest | FlagPingReply | FlagHello | FlagRumble | FlagAck | FlagReject | FlagConfig;


    public const int OffMagic = 0;
    public const int OffVersion = 2;
    public const int OffFlags = 3;
    public const int OffSeq = 4;
    public const int OffTimestamp = 6;
    public const int OffButtonsLo = 10;
    public const int OffButtonsHi = 12;
    public const int OffAxes = 14;

    public const byte RejectVersionMismatch = 1;
    public const byte RejectWrongPin = 2;
    public const byte RejectLobbyFull = 3;

    public const int MaxSlots = 4;
    public const ushort RumbleFrameIntervalMs = 33;

    /// <summary>
    /// Silence after which a slot's inputs are neutralized. 300 ms was too tight: on 2.4 GHz Wi-Fi a
    /// single retransmit burst or a background AP scan swallows ~5 snapshots and 2 keeper pings, and
    /// every spurious trip slams the wheel to centre and lifts every held key — in game that reads as
    /// steering stutter and a wheel that will not hold full lock. 800 ms clears four keeper intervals
    /// while still bounded well under the 2000 ms slot-free grace period.
    /// </summary>
    public const int FailsafeMs = 800;
}
